### PR TITLE
Replace `tendermint` with `cometbft`

### PR DESCRIPTION
**Pull request by Ferret-san**
use commetbft types in rollkit in order to comply with cosmos-sdk 0.47.0 by replacing tendermint with cometbft, and then replacing cometbft with the rollkit fork of tendermint

### CODE REVIEW

File: docs/lazy-adr/adr-002-mempool.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ For now, mempool implementation from lazyledger-core/Tendermint will be used.
 
 * inherit all limitations of the tendermint mempool
   * no prioritization of Txs
-  * many [open issues](https://github.com/tendermint/tendermint/issues?q=is%3Aissue+is%3Aopen+mempool+label%3AC%3Amempool)
+  * many [open issues](https://github.com/cometbft/cometbft/issues?q=is%3Aissue+is%3Aopen+mempool+label%3AC%3Amempool)
```
Comment (tzdybal) [inline docs/lazy-adr/adr-002-mempool.md:18]: I would leave this as is (without changing to cometbft).

If we're willing to re-import latest & greatest mempool from combetbft, we should do this in separate PR.

File: go.mod [hunk 2]
```diff
@@ -14,43 +15,55 @@ require (
...
 	google.golang.org/protobuf v1.30.0
 )
 
 require (
-	github.com/DataDog/zstd v1.4.5 // indirect
+	github.com/libp2p/go-libp2p-kad-dht v0.21.1
+	github.com/libp2p/go-libp2p-pubsub v0.9.3
+)
+
+require (
```
Comment (tzdybal) [inline go.mod:36]: Why? Maybe `go mod tidy` will fix this?

File: go.mod [hunk 3]
```diff
@@ -1,10 +1,11 @@
 module github.com/rollkit/rollkit
 
-go 1.18
+go 1.20
```
Comment (tzdybal) [inline go.mod:3]: :muscle:

File: node/full_client.go [hunk 4]
```diff
@@ -7,24 +7,25 @@ import (
...
-	"github.com/tendermint/tendermint/proxy"
-	rpcclient "github.com/tendermint/tendermint/rpc/client"
-	ctypes "github.com/tendermint/tendermint/rpc/core/types"
-	tmtypes "github.com/tendermint/tendermint/types"
-	"github.com/tendermint/tendermint/version"
+	abcicli "github.com/cometbft/cometbft/abci/client"
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cometbft/cometbft/config"
+	bytes "github.com/cometbft/cometbft/libs/bytes"
+	tmbytes "github.com/cometbft/cometbft/libs/bytes"
```
Comment (tzdybal) [inline node/full_client.go:14]: `cbbytes`? `cbftbytes`? `cometbytes`?

File: node/full_client_test.go [hunk 5]
```diff
@@ -813,8 +813,8 @@ func indexBlocks(t *testing.T, rpc *FullClient, heights []int64) {
 						Type: "begin_event",
 						Attributes: []abci.EventAttribute{
 							{
-								Key:   []byte("proposer"),
-								Value: []byte("FCAA001"),
+								Key:   string([]byte("proposer")),
+								Value: string([]byte("FCAA001")),
```
Comment (tzdybal) [inline node/full_client_test.go:817]: Keep it simple :wink: 
```suggestion
								Key:   "proposer",
								Value: "FCAA001",
```

File: node/full_client_test.go [hunk 6]
```diff
@@ -827,8 +827,8 @@ func indexBlocks(t *testing.T, rpc *FullClient, heights []int64) {
 						Type: "end_event",
 						Attributes: []abci.EventAttribute{
 							{
-								Key:   []byte("foo"),
-								Value: []byte(fmt.Sprintf("%d", h)),
+								Key:   string([]byte("foo")),
+								Value: string([]byte(fmt.Sprintf("%d", h))),
 								Index: true,
```
Comment (tzdybal) [inline node/full_client_test.go:832]: Encore
```suggestion
								Key:   "foo",
								Value: fmt.Sprintf("%d", h),
								Index: true,
```

File: proto/tendermint/abci/types.proto [hunk 7]
```diff
@@ -36,10 +35,13 @@ message Request {
     RequestOfferSnapshot      offer_snapshot       = 13;
     RequestLoadSnapshotChunk  load_snapshot_chunk  = 14;
     RequestApplySnapshotChunk apply_snapshot_chunk = 15;
-    RequestGetAppHash         get_app_hash         = 16;
-    RequestGenerateFraudProof generate_fraud_proof = 17;
-    RequestVerifyFraudProof   verify_fraud_proof   = 18;
+    RequestPrepareProposal    prepare_proposal     = 16;
+    RequestProcessProposal    process_proposal     = 17;
```
Comment (tzdybal) [inline proto/tendermint/abci/types.proto:39]: ++ :rocket:

File: proto/tendermint/abci/types.proto [hunk 8]
```diff
@@ -52,22 +54,17 @@ message RequestInfo {
...
 message RequestInitChain {
   google.protobuf.Timestamp time = 1
       [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
-  string                   chain_id         = 2;
-  ConsensusParams          consensus_params = 3;
-  repeated ValidatorUpdate validators       = 4 [(gogoproto.nullable) = false];
-  bytes                    app_state_bytes  = 5;
-  int64                    initial_height   = 6;
+  string                           chain_id         = 2;
+  tendermint.types.ConsensusParams consensus_params = 3;
```
Comment (tzdybal) [inline proto/tendermint/abci/types.proto:64]: Why `tendermint` is introduced / explicitly stated here?

File: proto/tendermint/abci/types.proto [hunk 9]
```diff
@@ -191,18 +218,10 @@ message ResponseInfo {
...
-  // bytes data = 2;
-  string log  = 3;
-  string info = 4;
-}
-
 message ResponseInitChain {
-  ConsensusParams          consensus_params = 1;
-  repeated ValidatorUpdate validators       = 2 [(gogoproto.nullable) = false];
-  bytes                    app_hash         = 3;
+  tendermint.types.ConsensusParams consensus_params = 1;
```
Comment (tzdybal) [inline proto/tendermint/abci/types.proto:222]: Again, this is surprising.

File: proto/tendermint/abci/types.proto [hunk 10]
```diff
@@ -256,9 +275,9 @@ message ResponseDeliverTx {
 }
 
 message ResponseEndBlock {
-  repeated ValidatorUpdate validator_updates       = 1 [(gogoproto.nullable) = false];
-  ConsensusParams          consensus_param_updates = 2;
-  repeated Event           events                  = 3
+  repeated ValidatorUpdate         validator_updates       = 1 [(gogoproto.nullable) = false];
+  tendermint.types.ConsensusParams consensus_param_updates = 2;
```
Comment (tzdybal) [inline proto/tendermint/abci/types.proto:279]: One more.

File: state/executor_test.go [hunk 11]
```diff
@@ -32,16 +34,22 @@ func doTestCreateBlock(t *testing.T, fraudProofsEnabled bool) {
 	app := &mocks.Application{}
 	app.On("CheckTx", mock.Anything).Return(abci.ResponseCheckTx{})
 
+	fmt.Println("App On CheckTx")
```
Comment (tzdybal) [inline state/executor_test.go:37]:
```suggestion
```

File: state/executor_test.go [hunk 12]
```diff
@@ -3,19 +3,21 @@ package state
 import (
 	"context"
 	"crypto/rand"
+	"fmt"
```
Comment (tzdybal) [inline state/executor_test.go:6]:
```suggestion
```

File: state/executor_test.go [hunk 13]
```diff
@@ -32,16 +34,22 @@ func doTestCreateBlock(t *testing.T, fraudProofsEnabled bool) {
 	app := &mocks.Application{}
 	app.On("CheckTx", mock.Anything).Return(abci.ResponseCheckTx{})
 
+	fmt.Println("App On CheckTx")
 	client, err := proxy.NewLocalClientCreator(app).NewABCIClient()
+	fmt.Println("Created New Local Client")
```
Comment (tzdybal) [inline state/executor_test.go:39]:
```suggestion
```

File: state/executor_test.go [hunk 14]
```diff
@@ -32,16 +34,22 @@ func doTestCreateBlock(t *testing.T, fraudProofsEnabled bool) {
...
 	require.NoError(err)
 	require.NotNil(client)
 
 	nsID := [8]byte{1, 2, 3, 4, 5, 6, 7, 8}
-
-	mpool := mempoolv1.NewTxMempool(logger, cfg.DefaultMempoolConfig(), proxy.NewAppConnMempool(client), 0)
-	executor := NewBlockExecutor([]byte("test address"), nsID, "test", mpool, proxy.NewAppConnConsensus(client), fraudProofsEnabled, nil, logger)
+	fmt.Println("Made NID")
```
Comment (tzdybal) [inline state/executor_test.go:44]:
```suggestion
```

File: state/executor_test.go [hunk 15]
```diff
@@ -32,16 +34,22 @@ func doTestCreateBlock(t *testing.T, fraudProofsEnabled bool) {
...
+	mpool := mempoolv1.NewTxMempool(logger, cfg.DefaultMempoolConfig(), proxy.NewAppConnMempool(client, proxy.NopMetrics()), 0)
+	fmt.Println("Made a NewTxMempool")
```
Comment (tzdybal) [inline state/executor_test.go:46]:
```suggestion
```

File: state/executor_test.go [hunk 16]
```diff
@@ -32,16 +34,22 @@ func doTestCreateBlock(t *testing.T, fraudProofsEnabled bool) {
...
+	executor := NewBlockExecutor([]byte("test address"), nsID, "test", mpool, proxy.NewAppConnConsensus(client, proxy.NopMetrics()), fraudProofsEnabled, nil, logger)
+	fmt.Println("Made a New Block Executor")
```
Comment (tzdybal) [inline state/executor_test.go:48]:
```suggestion
```

File: state/indexer/block/kv/kv_test.go [hunk 17]
```diff
@@ -30,8 +30,8 @@ func TestBlockIndexer(t *testing.T) {
 					Type: "begin_event",
 					Attributes: []abci.EventAttribute{
 						{
-							Key:   []byte("proposer"),
-							Value: []byte("FCAA001"),
+							Key:   string([]byte("proposer")),
+							Value: string([]byte("FCAA001")),
```
Comment (tzdybal) [inline state/indexer/block/kv/kv_test.go:34]:
```suggestion
							Key:   "proposer",
							Value: "FCAA001",
```

File: state/indexer/block/kv/kv_test.go [hunk 18]
```diff
@@ -44,8 +44,8 @@ func TestBlockIndexer(t *testing.T) {
 					Type: "end_event",
 					Attributes: []abci.EventAttribute{
 						{
-							Key:   []byte("foo"),
-							Value: []byte("100"),
+							Key:   string([]byte("foo")),
+							Value: string([]byte("100")),
```
Comment (tzdybal) [inline state/indexer/block/kv/kv_test.go:48]:
```suggestion
							Key:   "foo",
							Value: "100",
```

File: state/indexer/block/kv/kv_test.go [hunk 19]
```diff
@@ -68,8 +68,8 @@ func TestBlockIndexer(t *testing.T) {
 						Type: "begin_event",
 						Attributes: []abci.EventAttribute{
 							{
-								Key:   []byte("proposer"),
-								Value: []byte("FCAA001"),
+								Key:   string([]byte("proposer")),
+								Value: string([]byte("FCAA001")),
```
Comment (tzdybal) [inline state/indexer/block/kv/kv_test.go:72]:
```suggestion
								Key:   "proposer",
								Value: "FCAA001",
```

File: state/indexer/block/kv/kv_test.go [hunk 20]
```diff
@@ -82,8 +82,8 @@ func TestBlockIndexer(t *testing.T) {
 						Type: "end_event",
 						Attributes: []abci.EventAttribute{
 							{
-								Key:   []byte("foo"),
-								Value: []byte(fmt.Sprintf("%d", i)),
+								Key:   string([]byte("foo")),
+								Value: string([]byte(fmt.Sprintf("%d", i))),
```
Comment (tzdybal) [inline state/indexer/block/kv/kv_test.go:86]:
```suggestion
								Key:   "foo",
								Value: fmt.Sprintf("%d", i),
```

File: state/txindex/kv/kv_bench_test.go [hunk 21]
```diff
@@ -32,8 +32,8 @@ func BenchmarkTxSearch(b *testing.B) {
 			{
 				Type: "transfer",
 				Attributes: []abci.EventAttribute{
-					{Key: []byte("address"), Value: []byte(fmt.Sprintf("address_%d", i%100)), Index: true},
-					{Key: []byte("amount"), Value: []byte("50"), Index: true},
+					{Key: string([]byte("address")), Value: string([]byte(fmt.Sprintf("address_%d", i%100))), Index: true},
+					{Key: string([]byte("amount")), Value: string([]byte("50")), Index: true},
```
Comment (tzdybal) [inline state/txindex/kv/kv_bench_test.go:36]:
```suggestion
					{Key: "address", Value: fmt.Sprintf("address_%d", i%100), Index: true},
					{Key: "amount", Value: "50", Index: true},
```

File: state/txindex/kv/kv_test.go [hunk 22]
```diff
@@ -74,9 +74,9 @@ func TestTxSearch(t *testing.T) {
 	indexer := NewTxIndex(context.Background(), kvStore)
 
 	txResult := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("1"), Index: true}}},
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("owner"), Value: []byte("Ivan"), Index: true}}},
-		{Type: "", Attributes: []abci.EventAttribute{{Key: []byte("not_allowed"), Value: []byte("Vlad"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("1")), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("owner")), Value: string([]byte("Ivan")), Index: true}}},
+		{Type: "", Attributes: []abci.EventAttribute{{Key: string([]byte("not_allowed")), Value: string([]byte("Vlad")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:79]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 23]
```diff
@@ -149,9 +149,9 @@ func TestTxSearchWithCancelation(t *testing.T) {
...
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:154]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 24]
```diff
@@ -170,7 +170,7 @@ func TestTxSearchDeprecatedIndexing(t *testing.T) {
 
 	// index tx using events indexing (composite key)
 	txResult1 := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("1"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("1")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:173]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 25]
```diff
@@ -249,8 +249,8 @@ func TestTxSearchOneTxWithMultipleSameTagsButDifferentValues(t *testing.T) {
 	indexer := NewTxIndex(context.Background(), kvStore)
 
 	txResult := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("1"), Index: true}}},
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("2"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("1")), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("2")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:253]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 26]
```diff
@@ -273,7 +273,7 @@ func TestTxSearchMultipleTxs(t *testing.T) {
 
 	// indexed first, but bigger height (to test the order of transactions)
 	txResult := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("1"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("1")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:276]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 27]
```diff
@@ -284,7 +284,7 @@ func TestTxSearchMultipleTxs(t *testing.T) {
 
 	// indexed second, but smaller height (to test the order of transactions)
 	txResult2 := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number"), Value: []byte("2"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number")), Value: string([]byte("2")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:287]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 28]
```diff
@@ -304,9 +304,9 @@ func TestTxSearchMultipleTxs(t *testing.T) {
 	require.NoError(t, err)
 
 	// indexed fourth (to test we don't include txs with similar events)
-	// https://github.com/tendermint/tendermint/issues/2908
+	// https://github.com/cometbft/cometbft/issues/2908
 	txResult4 := txResultWithEvents([]abci.Event{
-		{Type: "account", Attributes: []abci.EventAttribute{{Key: []byte("number.id"), Value: []byte("1"), Index: true}}},
+		{Type: "account", Attributes: []abci.EventAttribute{{Key: string([]byte("number.id")), Value: string([]byte("1")), Index: true}}},
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:309]: Simplify `string([[byte(`.

File: state/txindex/kv/kv_test.go [hunk 29]
```diff
@@ -304,9 +304,9 @@ func TestTxSearchMultipleTxs(t *testing.T) {
 	require.NoError(t, err)
 
 	// indexed fourth (to test we don't include txs with similar events)
-	// https://github.com/tendermint/tendermint/issues/2908
+	// https://github.com/cometbft/cometbft/issues/2908
```
Comment (tzdybal) [inline state/txindex/kv/kv_test.go:307]: This new link doesn't work.

File: store/store_test.go [hunk 30]
```diff
@@ -162,17 +163,17 @@ func TestBlockResponses(t *testing.T) {
 			Events: []abcitypes.Event{{
 				Type: "test",
 				Attributes: []abcitypes.EventAttribute{{
-					Key:   []byte("foo"),
-					Value: []byte("bar"),
+					Key:   string([]byte("foo")),
+					Value: string([]byte("bar")),
```
Comment (tzdybal) [inline store/store_test.go:167]: Simplify!

File: types/pb/tendermint/abci/types.pb.go
Comment (tzdybal) [inline types/pb/tendermint/abci/types.pb.go:1]: I remember there was some discussion about this... :thinking:

**Review by tzdybal — CHANGES_REQUESTED** [review]
Few things:
1. Consider re-importing latest version of mempool from cometbft (as separate, follow-up PR).
2. `tm` prefix for imports should be changed - maybe `comet` would do the trick? `cometbytes`? `cometmath`? or `cb` - `cbbytes`, cbmath`, `cbquery`, etc.
3. Do not use `string([]byte("string"))`. Even in tests.

File: types/pb/tendermint/abci/types.pb.go
Comment (nashqueue) [reply types/pb/tendermint/abci/types.pb.go:1]: https://github.com/rollkit/rollkit/pull/769

File: go.mod (hunk 2 above)
Comment (Ferret-san) [reply go.mod:36]: sorry, what is this referring to, the require with the `// inderect`s, or the require above that?

File: node/full_client.go (hunk 4 above)
Comment (Ferret-san) [reply node/full_client.go:14]: yeah need to update the import names

File: node/full_client_test.go (hunk 5 above)
Comment (Ferret-san) [reply node/full_client_test.go:817]: oops, apologies for this lol

**Comment by Ferret-san** [comment]
> Few things:
> 
> 1. Consider re-importing latest version of mempool from cometbft (as separate, follow-up PR).
> 2. `tm` prefix for imports should be changed - maybe `comet` would do the trick? `cometbytes`? `cometmath`? or `cb` - `cbbytes`, cbmath`, `cbquery`, etc.
> 3. Do not use `string([]byte("string"))`. Even in tests.

1. Could you elaborate on what you mean by re-importing the latest version of mempool as a follow up PR?
2. On it
3. On it

**Comment by Wondertan** [comment]
I wonder if this can be a quick "dependency hell" fix as long as the app doesn't migrate to CometBFT.
We should try depending on celestia-node after this PR, to see if it works.
If works, it would make Rollkit being able to migrate to RPC/API much earlier and even use BlobModule right after it's done.

cc @nashqueue

**Comment by Ferret-san** [comment]
CometBFT is needed because of ABCI though

**Comment by Wondertan** [comment]
Yep, my point is that it can also bring additional benefit

**Comment by nashqueue** [comment]
> I wonder if this can be a quick "dependency hell" fix as long as the app doesn't migrate to CometBFT.

Core is already on comet bft https://github.com/celestiaorg/celestia-core/pull/980

**Comment by Wondertan** [comment]
> Core is already on comet bft https://github.com/celestiaorg/celestia-core/pull/980

It just pulls the changes; they didn't rename the module [here](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/go.mod#L1). Node imports it as tendermint, not comet. And if you migrate to comet, then there will be no conflicts between comet and tendermint

**Comment by Ferret-san** [comment]
need to resolve conflicts

File: da/test/da_test.go [hunk 31]
```diff
@@ -140,6 +149,38 @@ func TestRetrieve(t *testing.T) {
 	}
 }
 
+func startMockGRPCServ() *grpc.Server {
```
Comment (gupadhyaya) [inline da/test/da_test.go:152]: we should leave `startMockGRPCServ` and `startMockCelestiaNodeServer` in `da/test/da_test_helpers.go`

File: state/executor.go [hunk 32]
```diff
@@ -150,17 +150,22 @@ func (e *BlockExecutor) ApplyBlock(ctx context.Context, state types.State, block
 	}
 
 	abciValUpdates := resp.EndBlock.ValidatorUpdates
-	err = validateValidatorUpdates(abciValUpdates, state.ConsensusParams.Validator)
+	protoState, err := state.ToProto()
```
Comment (gupadhyaya) [inline state/executor.go:153]: why do you need this change? `state.ConsensusParams.Validator` does not work?

Comment (Ferret-san) [reply state/executor.go:153]: huh, my vs code complained about it when I changed it, but you are right now that I saw it, `state.Consensus.Validator` works

**Review by tzdybal — DISMISSED** [review]

**Review by tzdybal — DISMISSED** [review]
LGTM

**Review by tzdybal — APPROVED** [review]

**Review by gupadhyaya — APPROVED** [review]
